python/tvm/target/target.py

```diff
@@ -324,6 +324,31 @@ def arm_cpu(model="unknown", options=None):
     return Target(" ".join(["llvm"] + opts))
 
 
+def stm32mp1(options=None):
+    """Return stm32mp1 target and the associated cross compiler based on the SDK.
+
+    Parameters
+    ----------
+    options : str or list of str
+        Additional options
+    """
+
+    from tvm.contrib import cc
+
+    if "TVM_STM32MP_SDK_PATH" not in os.environ or  os.getenv("TVM_STM32MP_SDK_PATH") == "":
+        raise RuntimeError("Require environment variable TVM_STM32MP_SDK_PATH providing the SDK path directory\n"
+                           "example: export TVM_STM32MP_SDK_PATH=<your path>/STM32MP15-Ecosystem-v2.1.0/Developer-Package/SDK")
+
+    sdk_path = os.environ["TVM_STM32MP_SDK_PATH"]
+    cc = cc.cross_compiler(sdk_path + '/sysroots/x86_64-ostl_sdk-linux/usr/bin/arm-ostl-linux-gnueabi/arm-ostl-linux-gnueabi-g++',
+                           options=['-mthumb',
+                                    '-mfpu=neon-vfpv4',
+                                    '-mfloat-abi=hard',
+                                    '-mcpu=cortex-a7',
+                                    '--sysroot=' + sdk_path + '/sysroots/cortexa7t2hf-neon-vfpv4-ostl-linux-gnueabi'])
+
+    return arm_cpu("stm32mp1", options), cc
```

Comment:
for functions that live in this module, convention is to just return the target rather than `(target, cross_compiler)`. i'd like to see more about how this is used and perhaps I could suggest an alternate location to place the cross-compiler config.

Reply:
Hi @areusch I was questioning myself for the best way to achieve the STM32MP1 SDK path integration to get a cross-compiler configuration.
The STM32MP1 MPU from STMicroelectronics is provided with and OpenSTLinux distribution base on Yocto build system.
I can not afford to embed the full armv7 gcc tool chain in the distribution and therefore can not relink on the target the object file generated by TVM. 
So my goal is to build on the host the shared library itself so that I can load it through RPC on the target without the need to relink.
That why I would like to create an helper to get the cross-compiler based on the OpenSTLinux SDK path. The things are working well but it is maybe done at the wrong place.
First of all I was thinking to add this fonctionnality in python/tvm/contrib/stm32mp.py but I changed my mind.

Please advise and I will update accordingly.

BR
Vincent

Reply:
@junrushao1994 @zxybazh @areusch

Any update regarding where this patch could land?

Thanks
Vincent

Reply:
hi @vinceab sorry for the delay here. my main concern with this PR is how we can test it and document which version of the SDK it works with so we can ensure it continues to work after it is merged. I think the best we could do in the TVM CI is assert that the compilation flow works--we don't have any non-cloud hardware in the loop for the TVM CI, so we couldn't assert against real hardware. even with this level of testing though, i'm not sure the approach would scale to arbitrary cross-compiler.

i'm wondering if it might make sense to invert the dependency here--would it be better to put this kind of convenience function in a separate git repo, since it depends only on public TVM APIs? It would also be easier to organize related things such as examples.

I do think if we take this approach, it would be helpful for us to consider how we could deal with the discoverability problem (I.e. could we index or link to such repos from a TVM site). it would be helpful to discuss this in the discuss.tvm.ai forum as well.

Reply:
Hi, thanks @vinceab for the patch and I shared the same concern as @areusch. I think it's worth discussing on the forum to figure out the best way to organize all kinds of corss compiler for ad hoc hardware settings. We have to make sure the code we merged into tvm is maintainable and testable with current hardware resource.

Reply:
Hi @areusch @zxybazh,

I understand your point.
What I can propose at least is to provide an example on how to use STM32MP1 with current TVM API by adding the SDK cross compiler function into this example (It was my first though but I wanted to go one step further by integrating this mechanism inside the TVM python code because I spent some week to understand how I can generate a .so library compatible with my target).
Then I can open a discussion on how cross compiler for ad hoc settings could easily supported.

BR
Vincent

Reply:
@vinceab it would be great to better understand your pain points. i'm still concerned about checking in untestable code to python/, particularly in target.py. but perhaps if we have a more complete example it would help us to understand the shortcomings with our existing docs. in particular, so far i'm wondering whether we should have some type of wiki or page where implementers on a particular device could provide snippets such as these and discuss any wrinkles or problems they had.

cc @hogepodge

Reply:
This doesn't feel like not something I'm comfortable checking in as is as it encapsulates something beyond what you would find in the most standard cross-compiler toolchains in the stack. 

Ramana

Reply:
@areusch following different documentation on tvm.apache.org/docs it was not describe how to execute a relay object on a target that dos not have the toolchain embedded in its rootfs (most of device with small memory footprint does not embed its gcc toolchain).
I have to dig to understand that a shared library is sufficient.
The existing doc is focus on device like Rpi or desktop but for other kind of device, it is not straight forward as it should be.

I guess that the extension of the documentation could be sufficient to explain how to do with device like STM32MP1.
Maybe having also a python script example integrated into the TVM git could be the solution has it is for the other target like RPi.

Vincent

Reply:
@hogepodge is working on a fairly extensive refactor of the TVM tutorials--I think we should ensure there is a place for stuff like this. it would still be helpful to see your tutorial or perhaps the python scripts you're using with this code! perhaps there are other pieces we are missing or should checkin, or perhaps there is a particular way we should structure our docs to make this use case more straightforward.

Reply:
Hi @areusch,

Sorry for the long answer delay... here is the Python script I use to test TVM on STM32MP1.
 
[stm32mp15x_deploy_mobilenet-v1_0.5_128_quant.zip](https://github.com/apache/tvm/files/6350671/stm32mp15x_deploy_mobilenet-v1_0.5_128_quant.zip)